### PR TITLE
Pack Project struct to reduce gas costs

### PR DESCRIPTION
**Pull request by ryley-o**
Pack Project struct where possible to reduce gas costs

## Gas impacts
Mint costs are slightly reduced. Representative minter cost updates are shown in the table below.

_All gas costs assume project 3 mints avg(501-516) of 1000_
| Gas Test | Previous gas used to purchase | New gas used to purchase | % change |
| --- | --- | --- | --- |
| MinterSetPriceV2_V3Core | 146345 | 142557 | -2.6% (cheaper) |
| MinterDAExpV2_V3Core | 158183 | 154395 | -2.4% (cheaper) |

## Audit impacts
- No interfaces are affected by this code
- Would still be good to Cc ongoing auditors of this update

### CODE REVIEW

File: contracts/GenArt721CoreV3.sol [hunk 1]
```diff
@@ -78,6 +79,12 @@ contract GenArt721CoreV3 is ERC721, Ownable, IGenArt721CoreContractV3 {
     IAdminACLV0 public adminACLContract;
 
     struct Project {
+        uint24 invocations;
+        uint24 maxInvocations;
+        uint24 scriptCount;
+        uint64 completedTimestamp;
```
Comment (jakerockland) [inline contracts/GenArt721CoreV3.sol:85]: nit: can we please make an inline comment clarifying that `64` bytes is plenty for the life of all of our existences and many generations (all of them) to come?

**Review by jakerockland — APPROVED** [review]
This LGTM.

On a related note, curious your thoughts on storing some of the `string` values as raw `bytes32` instead of `string` for similar byte packing/gas optimization reasons?

File: contracts/GenArt721CoreV3.sol (hunk 1 above)
Comment (ryley-o) [reply contracts/GenArt721CoreV3.sol:86]: 570 billion years 😄 updated!

Comment (jakerockland) [reply contracts/GenArt721CoreV3.sol:86]: Haha thank you.

**Comment by ryley-o** [comment]
>On a related note, curious your thoughts on storing some of the string values as raw bytes32 instead of string for similar byte packing/gas optimization reasons?

Packing will only save gas when combining multiple sub-32-byte items into the same 32-byte storage slot. So we would only pack if we used smaller bytes, like two bytes16 in a row.

We could definitely change any of the string parameters to any bytes size, but we would limit the length of the field. If, for example, we changed license from `string` to `bytes32`, the artist would be limited to roughly 31 characters (depending on the characters, some take >1 byte). Any time a `bytes32` is able to be used, there is an associated efficiency gain because strings have to use an additional storage slot to record the number of slots they use (since they are unbounded in size, essentially a bytes array).

tldr - if we think any of the fields that are currently strings can always be < ~30 characters, we could switch to reduce gas costs when an artist is populating the items. If we think we can limit several fields to < ~15 characters, we could pack those values into the same slot, potentially finding more gas reductions. These types of changes may impact the calls coming from the frontend.
